source/agora/common/BitMask.d

```diff
@@ -0,0 +1,192 @@
+    construction.
+    This type is created for use as the validators signing bitmask and any
+    changes should ensure that it does not compromise that use.
+
+*******************************************************************************/
+
+module agora.common.BitMask;
+
+public struct BitMask
+{
+    import std.algorithm;
+    import std.conv;
+    import std.format;
+    import std.range;
+
+    @safe:
+
+    /// Serialization
+    public void toString (scope void delegate (scope const char[]) @safe sink) const
+    {
+        formattedWrite(sink, "%s", iota(this.length).map!(i => this[i] ? "1" : "0").join(""));
+    }
+
+    /// Support for Vibe.d serialization
+    public string toString () const
+    {
+        string ret;
+        scope void delegate (scope const char[]) @safe sink = (scope v) { ret ~= v; };
+        this.toString(sink);
+        return ret;
+    }
+
+    pure:
+
+    /// Support for Vibe.d deserialization
+    public static BitMask fromString (scope const(char)[] str)
+    {
+        auto bitmask = BitMask(str.length);
+        str.enumerate.each!((i, c)
+        {
+            if (c == '1')
+                bitmask[i] = true;
+        });
+        return bitmask;
+    }
+
+    nothrow:
+
+    /// Count of active validators who are expected / allowed to sign the block
+    public size_t length;
+
+    /// Bytes to hold bits to indicate if a validator has signed
+    private ubyte[] bytes;
+
+    public this (size_t length) inout
+    {
+        this.length = length;
+        if (length > 0)
+            this.bytes = new inout(ubyte)[1 + ((length - 1) / 8)];
+    }
+
+    public this (size_t length, in ubyte[] bytes)
+    {
+        this(length);
+        assert(this.bytes.length == bytes.length);
+        bytes.enumerate.each!((i, b) => this.bytes[i] = b);
+    }
+
+    // set the bits that are set in given BitMask
+    public auto opOpAssign (string op : "|") (in BitMask add)
+    {
+        assert(this.length == add.length, "BitMask assignment must be for same bit length");
+        iota(this.length).each!((size_t i)
+        {
+            if (add[i])
+                this[i] = true;
+        });
+        return this;
+    }
+
+    /// return the indices of bits set
+    public auto setIndices () const
+    {
+        return iota(this.length).filter!(i => this[i]);
+    }
+
+    /// return the indices of bits not set
+    public auto notSetIndices () const
+    {
+        return iota(this.length).filter!(i => !this[i]);
+    }
+
+    /// return the length of set bits
+    public size_t setCount () const
+    {
+        return this.setIndices.count!(i => this[i]);
+    }
+
+    /// Support for sorting by count of set bits not value
+    public int opCmp (in typeof(this) rhs) const
+    {
+        assert(this.length == rhs.length, "Comparing different sized BitMasks is not valid");
+        return this.setCount < rhs.setCount ? -1 : 1;
+    }
+
+    @nogc:
```

Comment:
can we explicitely declare @nogc on the functions instead? sometimes we re-arrange functions inside a file, and now the function's signature depends on where the function appears in the file

Reply:
I was following a style I had seen used in some other PRs recently. I am really not sure which is better. Position the functions based on attributes or explicitly specify them on each function. Any other opinions on this?

Reply:
hmm, I remember I used to see some Dlang code (outside our codebase), where I had to scroll through an entire file just to be able to see what the modifiers on a functions are

Reply:
Yeah I used it in Bitblob in the following case:
```D
struct BitBlob (size_t byteSize)
{
    @safe pure: // Everything
    this () {} // Constructor that `throw new`, so not `nothrow @nogc`
    string toString() {} // Allocates, obviously
    void toString (SinkType sink) {} // Does not allocate, but not provable. And can `throw new`
    nothrow @nogc: // Everything else
}
```

In other word, when the number of methods that don't get the attributes applied is very small, and we want the attribute to apply to everything, I sometimes use `attribute:`. But in this case (`@nogc`), it'd be better to have it on the methods for sure. The other attributes, I'm neutral.

Reply:
As the file is fairly long then I will put all the attributes on each method.